optuna/samplers/nsgaiii.py

```diff
@@ -0,0 +1,543 @@
+                    if i == len(population) - 1:
+                        population.pop()
+                    else:
+                        population[i] = population.pop()
+                    non_dominated_population.append(individual)
+                else:
+                    i += 1
+
+            for x in non_dominated_population:
+                for y in dominates_list[x.number]:
+                    dominated_count[y] -= 1
+
+            assert non_dominated_population
+            population_per_rank.append(non_dominated_population)
+
+        return population_per_rank
+
+    def after_trial(
+        self,
+        study: Study,
+        trial: FrozenTrial,
+        state: TrialState,
+        values: Optional[Sequence[float]],
+    ) -> None:
+        assert state in [TrialState.COMPLETE, TrialState.FAIL, TrialState.PRUNED]
+        if self._constraints_func is not None:
+            _process_constraints_after_trial(self._constraints_func, study, trial, state)
+        self._random_sampler.after_trial(study, trial, state, values)
+
+
+def multi_choose(n: int, k: int) -> int:
+    return math.factorial(n + k - 1) // math.factorial(k) // math.factorial(n - 1)
+
+
+def generate_default_reference_point(
+    objective_dimension: int, dividing_parameter: int = 3
+) -> np.ndarray:
+    """Generates default reference points which are `uniformly` spread on a hyperplane."""
+    reference_points = np.zeros(
+        (
+            multi_choose(objective_dimension, dividing_parameter),
+            objective_dimension,
+        )
+    )
+    for i, comb in enumerate(
+        itertools.combinations_with_replacement(range(objective_dimension), dividing_parameter)
+    ):
+        for j in comb:
+            reference_points[i, j] += 1.0
+    return reference_points
+
+
+def _normalize(
+    elite_population: List[FrozenTrial],
+    population: List[FrozenTrial],
+    weights: Optional[np.ndarray] = None,
+    method: str = "original",
+) -> np.ndarray:
+    """Normalizes objective values of population
+
+    An ideal point z* consists of minimums in each axis. Each objective value of population is
+    then subtracted by the ideal point.
+    An extreme point of each axis is (originally) defined as a minimum solution of ASF from the
+    population. After that, intercepts are calculate as intercepts of hyperplane which has all
+    the extreme points on it and used to rescale objective values.
+    """
+    # TODO(Shinichi) Propagate argument "weights" and "method" to the constructor
+
+    # Collect objective values
+    objective_dimension = len(population[0].values)
+    objective_matrix = np.zeros((len(elite_population + population), objective_dimension))
+    for i, trial in enumerate(elite_population + population):
+        objective_matrix[i] = np.array(trial.values, dtype=float)
+
+    # Subtract ideal point
+    objective_matrix -= np.min(objective_matrix, axis=0)
+    objective_matrix[objective_matrix < 10e-3] = 0.0
```

Comment:
Why this mask is needed?

Reply:
This is a kind of stabilization adopted from pymoo. I'll removed it until we realize why this mask is necesssary.